docs/docs/options.md
# 配置选项说明

## GUI 配置选项


## CLI 配置选项

### 指定工作文件夹路径 `-w <路径>` `--workspace <路径>`

指定工作文件夹路径。

### 指定 VS Code 路径 `--vscode <路径>`

指定 VS Code 可执行文件的路径。

### 指定编译器类型 `-s <类型>` `--setup <类型>`

指定要使用的编译器类型。可通过 `--help` 选项获取可用的类型名。

### 指定编译器路径 `-c <路径>` `--compiler <路径>`

指定编译器可执行文件的路径。

### 启用兼容模式 `-C` `--compat`

启用兼容模式。

### 生成快捷方式 `-d` `--desktop-shortcut`

仅 Windows 可用。生成指向工作区文件夹的桌面快捷方式。

### 指定语言标准 `-D <标准>` `--standard <标准>`

指定语言标准。若 CLI 中不提供，则根据编译器版本选取。

该选项的值形如 `c++98` `c++11` ... `c++23` ... `c89` `c99` ... `c17` ...。

### 不将编译器添加到 Path `-E` `--no-set-env`

仅在 MinGW 编译器下可用。若启用此开关，则不将编译器路径添加到 Path。

### 设置运行快捷键 `-H <快捷键>` `--run-hotkey <快捷键>`

设置运行快捷键。默认为 `f6`。

### 启用调试前检查 `-i` `--ascii-check`

仅在 GCC MinGW 编译器下可用。启用 MinGW 非 ASCII 文件名检查。

### 指定语言 `-L <c|cpp>` `--language <c|cpp>`

指定生成 C 还是 C++（默认）的配置。

### 配置完成后打开 VS Code `-o` `--open-vscode`

配置完成后打开 VS Code。

### 卸载多余的扩展 `-r` `--remove-extensions`

卸载多余的 VS Code 扩展。

### 不发送统计数据 `-S` `--no-stats`

不发送统计数据。

### 强制生成测试文件 `-t` `--test`

强制生成测试文件，即便工作文件夹中已经包含 `helloworld.{c,cpp}`。

### 禁用测试文件生成 `-T` `--no-test`

不生成测试文件。

## CLI 其它选项

### 显示版本信息 `-V` `--version`

显示版本信息并退出。

### 显示帮助信息 `-h` `--help`

显示本文档的简单版本并退出。按 CLI 选项的字典序排序。

### 启用非交互模式 `-y` `--assume-yes`

关闭命令行交互操作，启用非交互模式。

### 启用更多输出 `-v` `--verbose`

输出更详细的日志；可叠加使用，最多三次（如 `-vvv`）。

### 不输出 `q` `--quite`

不在标准输出输出任何日志信息。

### 指定日志路径 `-L <路径>` `--log-path <路径>`

指定日志文件路径。日志默认生成在程序的数据文件夹。使用 `-L /dev/null`（类 UNIX）或 `-L NUL`（Windows）禁用日志。

### 使用图形界面 `-g` `--gui`

当启用此开关时，将启动图形界面，且所有其它选项均被忽略，以下选项除外：
- 日志相关（`-v` `-q` `-L`）
- 跳过扩展管理（`--skip-ext-manage`）

### 启用 CLI `--cli`

此开关无作用。

### 跳过扩展管理 `--skip-ext-manage`

启用此开关后，扩展管理步骤将被跳过。此选项仅用于调试、诊断软件本身，不应在正常运行时使用。


CHANGELOG.md
# 更新日志

## v4.0.6
*2023.1.31*
- 修复了无法正确解析含非 UTF-8 输出的编译器版本信息的问题（[#23](https://github.com/VSCodeConfigHelper/v4/pull/23)）
- 更新了若干依赖库

## v4.0.5
*2022.11.21*
- 出于 Heroku 收费政策调整，禁用了 Heroku App 错误反馈

## v4.0.4
*2022.11.17*
- 修复了首次使用 MinGW 编译器时，无法正常运行测试程序的问题

## v4.0.3
*2022.10.14*
- 修复了在安装低版本 WebView 2 的情况下，程序无法正常运行的问题

## v4.0.2
*2022.9.6*
- Mac 下现允许使用 .app 路径作为 vscode 路径

## v4.0.1
*2022.9.4*
- 放宽了安装扩展的错误处理；即当安装扩展可能出现问题（进程未正常退出）时，仍然继续配置流程

## v4.0.0
*2022.9.3*
- 尝试修复了小概率无法安装扩展（即 `extension::install...` 失败）的问题
- 优化了配置中的图形界面显示

## v4.0.0-beta.15
*2022.9.1*
- 使用 Console Pauser 代替自定义脚本；增加 Mac 下 iTerm2 的支持
- 修复了 MSVC 不能生成 64 位可执行文件的问题
- 修复了 MSVC 的错误的问题匹配设置
- 对于非 MSVC 的构建任务，改用 `"type": "process"`
- 增加了更多命令行短选项
- 修复了无法安装 Apple Clang 的错误

## v4.0.0-beta.14
*2022.8.22*
- 修复了奇怪的无法在 Windows GUI 下启动 VS Code 的错误
- 修复了 CLI 参数存在错误时，Windows 下观察不到输出的问题

## v4.0.0-beta.13
*2022.8.22*
- 增加了 `--skip-ext-manage` 开关（GUI/CLI 皆有效），用于跳过可能导致错误的扩展管理步骤
- 增加了自定义运行快捷键的选项（全局设置）
- 再次修复了点击“新手模式”在非 Windows 系统上导致的错误
- 修复了“读取上次配置”功能失效的问题

## v4.0.0-beta.12
*2022.8.21*
- 默认不再显示警告。`-v` 现可叠加，以指定输出级别；增加了 `--quiet`
- CLI 现在可接受用户输入；即在运行程序时选择 VS Code、编译器和工作文件夹路径。
- Windows 下不带参数启动程序时，若 GUI 无法启动，则退化到 CLI
- 提示安装 MinGW 时，动态地获取最新版 MinGW 下载地址
- Windows 下的 CLI 不再附加到父进程控制台；改用新建控制台

## v4.0.0-beta.11
*2022.8.16*
- Windows 下检测 WebView 2 安装
- CLI 增加了日志路径选项

## v4.0.0-beta.10
*2022.8.14*
- 修复了点击“新手模式”在非 Windows 系统上导致的错误

## v4.0.0-beta.9
*2022.8.7*
- 修复了初次使用 beta 8 时无法启动的致命错误
- 优化了早期错误的反馈

## v4.0.0-beta.8
*2022.7.25*
- 增加安装编译器时的错误反馈
- 更改了日志路径，并不覆盖原有日志
- 修复了读取非 JSON 格式的 `keybindings.json` 时的运行时错误

## v4.0.0-beta.7
*2022.7.25*

- 优化 CLI 错误提示

## v4.0.0-beta.6
*2022.7.25*

- 升级到 Tauri 正式版
- 配置成功的统计计数功能（[#2](https://github.com/VSCodeConfigHelper/v4/issues/2)）
- 程序崩溃时发送日志
- 优化日志
- 调整代码结构

## v4.0.0-beta.5
*2022.5.13*

- 对于包含 `&` 和 ` ` 的工作文件夹路径给出警告。（Windows 暂停脚本在此情形下不工作。）

## v4.0.0-beta.4
*2022.5.9*

- 添加（恢复）根据编译器版本自动选择最新语言标准的功能。

## v4.0.0-beta.3
*2022.5.7*

- 添加捐助页面。

## v4.0.0-beta.2
*2022.4.22*

- 修复 CLI 启动参数的逻辑错误。

## v4.0.0-beta.1
*2022.4.5*

## v4.0.0-beta.0
*2022.4.5*

- 使用 Rust + Svelte 重构项目。
- 具体新增与弃用功能参见下表：

| 功能             | v1  | v2  | v3  | v4  |
| ---------------- | --- | --- | --- | --- |
| MinGW 安装       | ✔️   | ✔️   | ✔️   | ✔️   |
| MinGW 检测       | ❌   | ✔️   | ✔️   | ✔️   |
| VS Code 检测     | ❌   | ✔️   | ✔️   | ✔️   |
| 扩展安装         | ❔   | ✔️   | ✔️   | ✔️   |
| 工作文件夹验证   | ✔️   | ✔️   | ✔️   | ✔️   |
| 创建测试代码     | ✔️   | ✔️   | ✔️   | ✔️   |
| C 语言支持       | ✔️   | ✔️   | ✔️   | ✔️   |
| 常用编译参数     | ❌   | ❌   | ✔️   | ✔️   |
| 外部弹窗运行     | ❌   | ✔️   | ✔️   | ✔️   |
| “新手模式”       | ❌   | ❌   | ✔️   | ✔️   |
| MinGW 编译配置   | ✔️   | ✔️   | ✔️   | ✔️   |
| LLVM MinGW       | ❌   | ❌   | ❌   | ✔️   |
| MSVC 编译配置    | ❌   | ❌   | ❌   | ✔️   |
| Linux 支持       | ❌   | ❌   | ❔   | ✔️   |
| Linux GCC        | ❌   | ❌   | ✔️   | ✔️   |
| Linux Clang      | ❌   | ❌   | ❌   | ✔️   |
| macOS 支持       | ❌   | ❌   | ❔   | ✔️   |
| Apple Clang      | ❌   | ❌   | ✔️   | ✔️   |
| 非法调试前警告   | ❌   | ❌   | ✔️   | ✔️   |
| CLI              | ❌   | ❌   | ✔️   | ✔️   |
| 管理员配置       | ✔️   | ✔️   | ❌   | ❌   |
| Windows 快捷方式 | ❌   | ❌   | ✔️   | ✔️   |
| “离线”安装扩展   | ❌   | ❌   | ✔️   | ❌   |
| 安装中文语言包   | ❌   | ❌   | ✔️   | ❌   |

| 技术信息             | v1           | v2           | v3     | v4        |
| -------------------- | ------------ | ------------ | ------ | --------- |
| 主要语言             | C#           | C#           | C++    | Rust      |
| GUI                  | Windows Form | Windows Form | Vue.js | Svelte.js |
| 可脱机运行           | ✔️            | ✔️            | ❌      | ✔️         |
| 单文件应用           | ❌            | ❌            | ✔️      | ✔️         |
| 程序体积 （Windows） | 0.4M         | 0.5M         | 4.0M   | 7.7M     |
| 代码量（行）         | 600+         | 1600+        | 2000+  | 3800+     |

## [v3.x](https://github.com/VSCodeConfigHelper/v3/blob/main/CHANGELOG.md)

## [v2.x](https://github.com/VSCodeConfigHelper/v2/blob/v2.x/CHANGELOG.md)



src-tauri/src/cli.rs
// Copyright (C) 2022 Guyutongxue
//
// This file is part of vscch4.
//
// vscch4 is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// vscch4 is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with vscch4.  If not, see <http://www.gnu.org/licenses/>.

use std::str::FromStr;

#[allow(unused_imports)]
use ::log::{debug, error, info, warn};
use anyhow::{anyhow, Result};
use clap::{AppSettings, ArgEnum, CommandFactory, Parser};
use std::io::Write;

use crate::gui::gui;
use crate::log;
use crate::steps::compiler::{CompilerSetup, ENABLED_SETUPS};
use crate::steps::options::Options;
use crate::steps::{vscode, workspace};
use crate::tasks;

#[cfg(windows)]
use crate::utils::winapi;
use crate::utils::ToString;

#[derive(Parser)]
#[clap(
  global_setting(AppSettings::NoAutoVersion),
  global_setting(AppSettings::NoAutoHelp),
  author, version, about, long_about = None
)]
struct CliArgs {
  /// 显示此帮助信息并退出
  #[clap(short, long, exclusive = true)]
  help: bool,

  /// 显示程序版本信息并退出
  #[clap(short = 'V', long, exclusive = true)]
  version: bool,

  #[clap(flatten)]
  verbose: clap_verbosity_flag::Verbosity,

  /// 日志路径
  #[clap(long)]
  log_path: Option<String>,

  /// 关闭命令行交互操作，总是假设选择“是”
  #[clap(short = 'y', long)]
  assume_yes: bool,

  /// 使用图形界面。当不提供任何命令行参数时，将优先运行 GUI
  #[clap(short = 'g', long)]
  gui: bool,

  /// 使用命令行界面（此选项没有其它作用）
  #[clap(long)]
  cli: bool,

  /// 指定 VS Code 安装路径。若不提供，则尝试自动检测
  #[clap(long)]
  vscode: Option<String>,

  /// 指定编译器类型（见后）
  #[clap(short, long, default_value = DEFAULT_SETUP_ID)]
  setup: &'static CompilerSetup,

  /// 指定编译器路径。若不提供，则尝试寻找已安装的编译器
  #[clap(short, long)]
  compiler: Option<String>,

  /// 指定工作文件夹路径。使用 CLI 时必须提供
  #[clap(short, long)]
  workspace: Option<String>,

  /// 指定配置目标语言
  #[clap(short, long, arg_enum, default_value = "cpp")]
  language: Language,

  /// 指定语言标准。若不提供，则工具根据编译器版本选取
  #[clap(long, possible_values = ["c++98", "c++11", "c++14", "c++17", "c++20", "c++23", "c89", "c99", "c11", "c17"])]
  standard: Option<String>,

  /// 指定编译选项
  #[clap(short, long)]
  args: Vec<String>,

  /// 启用兼容模式
  #[clap(long)]
  compat: bool,

  /// 卸载多余的 VS Code 扩展
  #[clap(long)]
  remove_extensions: bool,

  /// 在调试前进行文件名中非 ASCII 字符的检查。仅 Windows 可用
  #[clap(long)]
  ascii_check: bool,

  /// 不将编译器添加到 Path 环境变量。仅 Windows 可用
  #[clap(long)]
  no_set_env: bool,

  /// 生成指向工作区文件夹的桌面快捷方式。仅 Windows 可用
  #[clap(long)]
  desktop_shortcut: bool,

  /// 配置完成后打开 VS Code
  #[clap(short, long)]
  open_vscode: bool,

  /// 强制生成测试文件
  #[clap(long, conflicts_with = "no-test")]
  test: bool,

  /// 不生成测试文件
  #[clap(long)]
  no_test: bool,

  /// 不发送统计数据
  #[clap(long)]
  no_stats: bool,
}

#[derive(Clone, PartialEq, ArgEnum)]
enum Language {
  #[clap(name = "cpp")]
  Cpp,
  #[clap(name = "c")]
  C,
}

static DEFAULT_SETUP_ID: &'static str = ENABLED_SETUPS[0].id;

impl FromStr for &'static CompilerSetup {
  type Err = anyhow::Error;
  fn from_str(s: &str) -> Result<Self, Self::Err> {
    ENABLED_SETUPS
      .iter()
      .find(|setup| setup.id == s)
      .cloned()
      .ok_or_else(|| anyhow!("Unknown compiler setup: {}", s))
  }
}

fn print_setup_help() {
  println!("\n\x1b[33mSETUPS:\x1b[0m");
  for setup in ENABLED_SETUPS {
    println!("    \x1b[32m{:10}\x1b[0m - {}", setup.id, setup.name);
  }
}

fn parse_args() -> Result<CliArgs> {
  let args = CliArgs::try_parse()?;
  log::setup(args.log_path.as_ref(), args.verbose.log_level_filter())?;
  tasks::statistics::set(!args.no_stats);
  Ok(args)
}

#[cfg(windows)]
fn has_webview2_installed() -> bool {
  use crate::utils::winreg;
  if let Some(v) = winreg::get(
    winreg::HKEY_LOCAL_MACHINE,
    r#"SOFTWARE\WOW6432Node\Microsoft\EdgeUpdate\Clients\{F3017226-FE2A-4295-8BDF-00C3A9A7E4C5}"#,
    "pv",
  ) {
    if v != "" && v != "0.0.0.0" {
      return true;
    }
  }
  if let Some(v) = winreg::get(
    winreg::HKEY_CURRENT_USER,
    r#"Software\Microsoft\EdgeUpdate\Clients\{F3017226-FE2A-4295-8BDF-00C3A9A7E4C5}"#,
    "pv",
  ) {
    if v != "" && v != "0.0.0.0" {
      return true;
    }
  }
  return false;
}

#[macro_export]
macro_rules! out {
  ($pattern:expr $(, $arg:expr)*) => (
    println!(concat!("\x1b[1m[OUT] ", $pattern, "\x1b[0m") $(, $arg)*)
  )
}

pub fn run() {
  let no_args = std::env::args().len() <= 1;

  let args = match parse_args() {
    Ok(args) => args,
    Err(e) => {
      if no_args {
        native_dialog::MessageDialog::new()
          .set_title("早期错误")
          .set_text(&format!("{:?}", e))
          .show_alert()
          .unwrap();
      } else {
        println!("[\x1b[31mERROR\x1b[0m] 早期错误：{:?}", e);
      }
      std::process::exit(1)
    }
  };

  if no_args {
    #[cfg(windows)]
    if !has_webview2_installed() {
      warn!("WebView 2 未安装。退化到命令行界面。");
      out!("检测到您的计算机上尚未安装 WebView 2；将使用命令行界面替代。建议您前往 https://go.microsoft.com/fwlink/p/?LinkId=2124703 下载 WebView 2 以获得更好体验。");
      cli_handled(args);
      return;
    }
    gui_handled();
    return;
  }

  if args.gui {
    gui_handled()
  } else {
    cli_handled(args)
  }
}

fn cli_handled(args: CliArgs) {
  match cli(args) {
    Ok(_) => (),
    Err(e) => {
      error!("{:?}", e);
      out!(
        "如果你认为该错误是 bug，请将 {} 文件发送到 guyutongxue@163.com。",
        log::get_log_path().to_string()
      )
    }
  }
  #[cfg(windows)]
  {
    out!("按任意键退出...");
    winapi::getch();
  }
}

fn gui_handled() {
  #[cfg(windows)]
  winapi::free_console();
  match gui() {
    Ok(_) => (),
    Err(e) => {
      if let Some(id) = tasks::statistics::send_error(&e) {
        native_dialog::MessageDialog::new()
          .set_title("程序已报告错误")
          .set_text(&format!(
            "{}\n您可以将代码 “{}” 发送至 guyutongxue@163.com，开发者会尽快帮您解决问题。",
            e.to_string(),
            &id[0..6]
          ))
          .set_type(native_dialog::MessageType::Error)
          .show_alert()
          .unwrap();
      }
    }
  }
}

fn cli(mut args: CliArgs) -> Result<()> {

  if args.help {
    CliArgs::command().print_help().unwrap();
    print_setup_help();
    return Ok(());
  } else if args.version {
    print!("{}", CliArgs::command().render_version());
    println!(
      r"Copyright (C) 2022 Guyutongxue
本程序是自由软件；你可以再分发之和/或依照由自由软件基金会发布的
GNU 通用公共许可证修改之，无论是版本 3 许可证，还是（按你的决定）
任何以后版都可以。本程序不含任何保障。"
    );
    return Ok(());
  }

  if !cfg!(windows) {
    fn nonsupport_check(name: &'static str, flag: &mut bool) {
      if *flag {
        warn!("{} 选项在此操作系统上不支持，已忽略。", name);
        *flag = false;
      }
    }
    nonsupport_check("--ascii-check", &mut args.ascii_check);
    nonsupport_check("--no-set-env", &mut args.no_set_env);
    nonsupport_check("--desktop-shortcut", &mut args.desktop_shortcut);
  }

  info!("验证 VS Code 安装...");
  let vscode = match args.vscode {
    Some(vscode) => {
      if let Err(e) = vscode::verify(&vscode) {
        return Err(anyhow!("VS Code 验证失败：{}", e));
      }
      vscode
    }
    None => {
      if args.assume_yes {
        match vscode::scan() {
          Some(v) => v,
          None => return Err(anyhow!("未找到 VS Code 安装。")),
        }
      } else {
        let mut builder = requestty::Question::input("vscode")
          .message("VS Code 路径：")
          .validate_on_key(|s, _| vscode::verify(s).is_ok())
          .validate(|s, _| vscode::verify(s).map_err(|s| format!("验证失败：{}", s)));
        if let Some(dft) = vscode::scan() {
          builder = builder.default(dft);
        }
        requestty::prompt_one(builder.build())?
          .as_string()
          .unwrap()
          .into()
      }
    }
  };

  info!("VS Code 安装在 {}，", vscode);

  info!("验证工作区路径...");
  let workspace = args
    .workspace
    .ok_or(anyhow!("未在命令行中指定工作区文件夹。"))?;
  if let Err(e) = workspace::path_available(&workspace) {
    return Err(anyhow!("工作文件夹验证失败: {}", e));
  }
  info!("工作区路径为 {}。", workspace);

  info!("验证 {} 类型编译器...", args.setup.id);
  let compiler = match args.compiler.as_ref() {
    // 验证命令行传入的编译器
    Some(compiler) => match args.setup.verify {
      Some(verify) => {
        verify(&compiler).map_err(|str| anyhow!("验证编译器 {} 失败：{}", &compiler, str))?
      }
      None => Err(anyhow!("该编译器类型不支持自定义。"))?,
    },
    // 寻找已安装的编译器
    None => {
      let compilers = (args.setup.scan)();
      debug!("找到的编译器有：{:?}", compilers);
      if compilers.len() == 0 {
        Err(anyhow!("未在命令行传入编译器，也找不到已安装的编译器。"))?
      } else {
        // TODO: selection
        compilers.into_iter().nth(0).unwrap()
      }
    }
  };
  info!("编译器为 {}。", compiler.path);

  let language = if args.language == Language::Cpp {
    "C++"
  } else {
    "C"
  };
  let add_to_path = !args.no_set_env;
  let test = if args.test {
    Some(true)
  } else if args.no_test {
    Some(false)
  } else {
    None
  };

  let task_init_args = tasks::TaskInitArgs {
    vscode: vscode,
    workspace: workspace,
    compiler: compiler,
    options: Options {
      language: language.into(),
      args: args.args,
      standard: args.standard,
      compatible_mode: args.compat,
      remove_extensions: args.remove_extensions,
      ascii_check: args.ascii_check,
      add_to_path: add_to_path,
      desktop_shortcut: args.desktop_shortcut,
      test: test,
      open_vscode: args.open_vscode,
      collect_data: !args.no_stats,
    },
  };
  debug!("task_init_args: {:?}", task_init_args);

  info!("正在初始化任务列表...");
  let task_list = tasks::list(task_init_args);
  debug!(
    "任务列表：{:?}",
    task_list.iter().map(|t| t.0).collect::<Vec<_>>()
  );

  for (name, action) in task_list {
    info!("正在执行任务 {}...", name);
    action()?;
    info!("任务 {} 执行完毕。", name);
  }
  Ok(())
}
